// @react-facet/dom-fiber/dist/testSetup.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.render = exports.act = exports.root = exports.setupAct = void 0;
const react_1 = __importDefault(require("react"));
const createFiberRoot_1 = require("./createFiberRoot");
const createReconciler_1 = require("./createReconciler");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getGlobalThis() {
    if (typeof globalThis !== 'undefined') {
        return globalThis;
    }
    if (typeof self !== 'undefined') {
        return self;
    }
    if (typeof window !== 'undefined') {
        return window;
    }
    if (typeof global !== 'undefined') {
        return global;
    }
    throw new Error('unable to locate global object');
}
/**
 * More info: https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#configuring-your-testing-environment
 * Reference implementation: https://github.com/testing-library/react-testing-library/blob/c80809a956b0b9f3289c4a6fa8b5e8cc72d6ef6d/src/act-compat.js#L5
 */
const setupAct = () => {
    getGlobalThis().IS_REACT_ACT_ENVIRONMENT = true;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const act = react_1.default.unstable_act;
    return act;
};
exports.setupAct = setupAct;
document.body.innerHTML = `<div id="root"></div><div id="portal"></div>`;
const reconcilerInstance = (0, createReconciler_1.createReconciler)();
exports.root = document.getElementById('root');
const fiberRoot = (0, createFiberRoot_1.createFiberRoot)(reconcilerInstance)(exports.root);
exports.act = (0, exports.setupAct)();
/**
 * Render function local to testing that shared the same instance of the reconciler.
 *
 * This is needed otherwise React complains that we are sharing a context across different renderers.
 */
const render = function render(ui) {
    (0, exports.act)(() => {
        reconcilerInstance.updateContainer(ui, fiberRoot, null, () => { });
    });
};
exports.render = render;
afterEach(() => {
    (0, exports.act)(() => {
        reconcilerInstance.updateContainer(null, fiberRoot, null, () => { });
    });
});
